### ISSUE TITLE
AbortSignal integration

### ISSUE THREAD

**Issue by rektide** [connor4312/cockatiel#46]
Hello, 

This library looks wonderful. One thing that makes it a bit hard for us & probably other folks to adopt & use is that it escew's the web (and increasingly the rest of js too) standard [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) cancellation protocols. 

There are some various api differences between the standard JS abort protocols & the [`CancellationToken`](https://github.com/connor4312/cockatiel/blob/90e865b2dbd4e086b348e1e7a79650c1c427ed84/readme.md#cancellationtoken) presently in use in cockatiel. The main difference is that `CancellationToken` is both the means to observe change, and the system for doing the change. This is reminisce of the old jQuery [`Deferrable`](https://api.jquery.com/category/deferred-object/), which was a precursor to promises, where the Deferrable was both the promise & the resolver. In JS today, usually the thing that does the actuating/changing is separate from the system for observing. That is reflected in `AbortSignal` and it's pair [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController), where `AbortSignal` is the output (somewhat alike a sychronously readable Promise), and `AbortController` can be used to actuate the signal (quite strongly resembling the obsolete JS [`Deferred`](http://web.archive.org/web/20201111210253/https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred), which contrary to jquery, divided the promise & the resolver ala AbortSignal/AbortController).  First used in `Fetch`, today, AbortSignal is being widely integrated into the JS ecosystem.

Please pardon my long-windedness above. I find it interesting to provide some historical roots for the topic, as they show how things have moved over time, but I am mainly writing this ticket to talk about today & the future & how cockatiel can be the best JS citizen it can be. [`CancellationToken`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken) has a pretty nice, convenient, & very powerful set of APIs atop it, and I generally commend the Microsoft teams that have shaped it over the years. However, I feel like it's in cockatiel's best interest prefer to integrate with the JS language ecosystem, if possible, and feel like we should try to switch to AbortSignal/AbortController as the primary means to handle/deal-with cancellation concerns.

**Comment by connor4312** [comment]
Yea, I think I want to move to AbortSignal in a v3 of this library.

**Comment by HonzaMac** [comment]
I tried to use this pattern to forward `new AbortController().signal` into `CancellationToken`. 
Could this work?
```
import fetch, {RequestInfo, RequestInit, Response} from 'node-fetch';

export const fetchRetry = (url: RequestInfo, init: RequestInit): Promise<Response> => {
  const cts = new CancellationTokenSource();
  if (init?.signal) {
    init.signal.addEventListener('abort', () => cts.cancel(), {once: true});
  }

  return retryPolicy.execute(() => fetch(url, init), cts.token);
};
```

**Comment by connor4312** [comment]
I've now done this. However, as AbortController/AbortToken is only available in Node.js starting in v16, I will wait to publish this until the end of April when Node 12 enters EOL and Node 18 is released.

**Comment by HonzaMac** [comment]
Can we use some of available Abort Controller polyfills? And replace it with native on April?

**Comment by connor4312** [comment]
I published `3.0.0-beta.0`. It's a prerelease tag so will need to be explicitly installed.

**Comment by HonzaMac** [comment]
Thank you! Will try it on mine project.

**Comment by dev-usa** [comment]
> I published `3.0.0-beta.0`. It's a prerelease tag so will need to be explicitly installed.

Trying to use this lib both the current version and the `3.0.0-beta.0` in a node application that I am working and receiving the AbortController is not defined error. Please see below. Tried using the [AbortController polyfill](https://github.com/mo/abortcontroller-polyfill) but could not get it to work. Any suggestions? I am running node v14.17.3

```shell
2022-01-05T19:28:56.943Z        undefined       ERROR   Uncaught Exception      {"errorType":"ReferenceError","errorMessage":"AbortController is not defined","stack":["ReferenceError: AbortController is not defined","   
 at Object.<anonymous> (/var/task/node_modules/cockatiel/dist/common/abort.js:7:34)","    
at Module._compile (internal/modules/cjs/loader.js:1085:14)","    
at Object..js (internal/modules/cjs/loader.js:1114:10)","    
at Module.load (internal/modules/cjs/loader.js:950:32)","    
at Function._load (internal/modules/cjs/loader.js:790:14)","    
at Module.require (internal/modules/cjs/loader.js:974:19)","    
at require (internal/modules/cjs/helpers.js:92:18)","    
at Object.<anonymous> (/var/task/node_modules/cockatiel/dist/CircuitBreakerPolicy.js:4:17)","    
at Module._compile (internal/modules/cjs/loader.js:1085:14)","    
at Object..js (internal/modules/cjs/loader.js:1114:10)"]}
time="2022-01-05T19:28:56.95" level=panic msg="ReplyStream not available"
```

**Comment by connor4312** [comment]
Yes, you must use it in an environment where AbortController is available, that's why it's a prerelease for now. Or define the AbortController in some way, I don't know anything about that polyfill or your use of it.
